### PR TITLE
[MRG] Sample weights in SAGA

### PR DESCRIPTION
**Pull request by fabianp**
Second implementation, as discussed in #68

I made some benchmarks to compare this (strategy 2) against the unweighted version (strategy 1):

image
image
image

The first column is a sanity check to make sure that they are indeed the same in terms of number of iterations (as it should be). The second column represent the convergence with respect to time. My interpretation is that the overhead is negligible and that any difference that we might be seeing is probably dominated by noise in the measurements (it doesnt make much sense that the weighted version is faster as it seems to be the case in the first dataset.

### CODE REVIEW

**Comment by mblondel** [comment]
Thanks a lot for the investigation. I prefer this design, if this is ok with you too.

**Comment by fabianp** [comment]
Yep, I also prefer it.

File: lightning/impl/tests/test_sag.py [hunk 1]
```diff
@@ -401,3 +401,20 @@ def test_sag_sparse():
                                    max_iter=1, random_state=0, alpha=alpha)
         clf_dense.fit(X.toarray(), y)
         assert_equal(clf_sparse.score(X, y), clf_dense.score(X, y))
+
+
+def test_sag_sample_weights():
```
Comment (mblondel) [inline lightning/impl/tests/test_sag.py]: Can you add a test with non-uniform weights?

**Comment by fabianp** [comment]
Fixed the tests and added an example, should be good to go.

**Comment by fabianp** [comment]
hey @mblondel , green light to merge this?

**Comment by mblondel** [comment]
Yes! Thanks!

**Pull request merged**